<a href="https://colab.research.google.com/github/kaancceylan/medium-notebooks/blob/main/Pandas_Medium_Post.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Pandas for Medium
Notebook for the medium post of [Essential Libraries To Have In Your Toolbox For Data Science And ML - Series #2 - Pandas](https://medium.com/@kaanceylan/essential-libraries-to-have-in-your-tool-for-data-science-and-ml-series-1-numpy-a7ca43f9a313?source=user_profile---------0-------------------------------)

Data from [Kaggle](https://www.kaggle.com/jsphyg/tipping)


Import pandas and read the dataset in from the csv file. 

In [ ]:
import pandas as pd
df = pd.read_csv('tips.csv')

You can use index_col and pass a column to set as the index of the dataframe if you want. For example, you can set days as the row labels of the dataframe and separate the data by days

In [ ]:
df2 = pd.read_csv('tips.csv', index_col = 'day')
df2.head()

total_bill   tip     sex smoker    time  size
day                                               
Sun       16.99  1.01  Female     No  Dinner     2
Sun       10.34  1.66    Male     No  Dinner     3
Sun       21.01  3.50    Male     No  Dinner     3
Sun       23.68  3.31    Male     No  Dinner     2
Sun       24.59  3.61  Female     No  Dinner     4

##Basic Functions

In [ ]:
df.shape

(244, 7)

In [ ]:
df.columns

Index(['total_bill', 'tip', 'sex', 'smoker', 'day', 'time', 'size'], dtype='object')

244 rows of 7 different features in the dataset. Column names are 
-  total_bill
-  tip
-  sex
-  smoker
-  day
-  time
-  size

Basic descriptive statistics on the dataset with describe()

In [ ]:
df.describe()

total_bill         tip        size
count  244.000000  244.000000  244.000000
mean    19.785943    2.998279    2.569672
std      8.902412    1.383638    0.951100
min      3.070000    1.000000    1.000000
25%     13.347500    2.000000    2.000000
50%     17.795000    2.900000    2.000000
75%     24.127500    3.562500    3.000000
max     50.810000   10.000000    6.000000

You can get the non-null values for each column, and the datatype that the column holds.

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 244 entries, 0 to 243
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   total_bill  244 non-null    float64
 1   tip         244 non-null    float64
 2   sex         244 non-null    object 
 3   smoker      244 non-null    object 
 4   day         244 non-null    object 
 5   time        244 non-null    object 
 6   size        244 non-null    int64  
dtypes: float64(2), int64(1), object(4)
memory usage: 13.5+ KB


Head prints the first 5 rows of the dataset by deault but you can pass different numbers to take a quick look at the data and see what it looks like.

In [ ]:
df.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

In [ ]:
df.dtypes

total_bill    float64
tip           float64
sex            object
smoker         object
day            object
time           object
size            int64
dtype: object

##Slicing and Filtering The Data

You can use loc and iloc to slice the part of the dataset that you want to process seperately or do calculations on.

Printing all the rows where the total_bill is greater than 24.

In [ ]:
df.loc[df['total_bill'] > 24]

total_bill   tip     sex smoker  day    time  size
4         24.59  3.61  Female     No  Sun  Dinner     4
5         25.29  4.71    Male     No  Sun  Dinner     4
7         26.88  3.12    Male     No  Sun  Dinner     4
11        35.26  5.00  Female     No  Sun  Dinner     4
23        39.42  7.58    Male     No  Sat  Dinner     4
..          ...   ...     ...    ...  ...     ...   ...
230       24.01  2.00    Male    Yes  Sat  Dinner     4
237       32.83  1.17    Male    Yes  Sat  Dinner     2
238       35.83  4.67  Female     No  Sat  Dinner     3
239       29.03  5.92    Male     No  Sat  Dinner     3
240       27.18  2.00  Female    Yes  Sat  Dinner     2

[64 rows x 7 columns]

Tip amount for groupsize > 6 and total_bill > 24

In [ ]:
df.loc[(df['total_bill'] > 24) & (df['size'] > 5), ['tip']]

tip
125  4.2
141  6.7
143  5.0
156  5.0

Can do the same thing with iloc but iloc does not accept the filtering we did with loc, iloc takes integers or boolean arrays and the list function inside the iloc returns an array of boolean values.

In [ ]:
print(list(df.total_bill > 24))

[False, False, False, False, True, True, False, True, False, False, False, True, False, False, False, False, False, False, False, False, False, False, False, True, False, False, False, False, False, False, False, False, False, False, False, True, False, False, False, True, False, False, False, False, True, False, False, True, True, False, False, False, True, False, True, False, True, True, False, True, False, False, False, False, False, False, False, False, False, False, False, False, True, True, False, False, False, True, False, False, False, False, False, True, False, True, False, False, True, False, True, False, False, False, False, True, True, False, False, False, False, False, True, False, False, False, False, True, False, False, False, False, True, False, True, False, True, False, False, True, False, False, False, False, False, True, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, True, True, True, False, False, False, Fals

In [ ]:
df_filter_size = df.iloc[list(df.total_bill > 24)]

In [ ]:
print(df_filter_size)

     total_bill   tip     sex smoker  day    time  size
4         24.59  3.61  Female     No  Sun  Dinner     4
5         25.29  4.71    Male     No  Sun  Dinner     4
7         26.88  3.12    Male     No  Sun  Dinner     4
11        35.26  5.00  Female     No  Sun  Dinner     4
23        39.42  7.58    Male     No  Sat  Dinner     4
..          ...   ...     ...    ...  ...     ...   ...
230       24.01  2.00    Male    Yes  Sat  Dinner     4
237       32.83  1.17    Male    Yes  Sat  Dinner     2
238       35.83  4.67  Female     No  Sat  Dinner     3
239       29.03  5.92    Male     No  Sat  Dinner     3
240       27.18  2.00  Female    Yes  Sat  Dinner     2

[64 rows x 7 columns]


We've got the same result that we got with loc before.
Let's say we want to see the group size distribution of these rows. We can use the value_counts() and pass the 'size' to get that and as you can see, 48 out of the 64 rows had a group size of 3 or more.

In [ ]:
df_filter_size.value_counts('size')

size
4    26
2    16
3    14
5     4
6     4
dtype: int64

By writing df['a filtering condition'], we can filter the results of the dataframe. The same as we did with iloc, pass a boolean array to the df by writing the filter inside the df[]. Let's filter the dataset by the smokers.

In [ ]:
df[df['smoker']=='Yes']

total_bill   tip     sex smoker  day    time  size
56        38.01  3.00    Male    Yes  Sat  Dinner     4
58        11.24  1.76    Male    Yes  Sat  Dinner     2
60        20.29  3.21    Male    Yes  Sat  Dinner     2
61        13.81  2.00    Male    Yes  Sat  Dinner     2
62        11.02  1.98    Male    Yes  Sat  Dinner     2
..          ...   ...     ...    ...  ...     ...   ...
234       15.53  3.00    Male    Yes  Sat  Dinner     2
236       12.60  1.00    Male    Yes  Sat  Dinner     2
237       32.83  1.17    Male    Yes  Sat  Dinner     2
240       27.18  2.00  Female    Yes  Sat  Dinner     2
241       22.67  2.00    Male    Yes  Sat  Dinner     2

[93 rows x 7 columns]

Let's see how much missing value we have in our dataset. For that we'll use the isna() (you can also use isnull() ) function along with the sum() function so that the results are easier the read. isna() detects all the missing values in the dataframe and sum() counts them, giving us the final result. If we use only isna(), we get the same dataframe but the values will be masked with False or True values based on if they're null or not.

In [ ]:
df.isna()

total_bill    tip    sex  smoker    day   time   size
0         False  False  False   False  False  False  False
1         False  False  False   False  False  False  False
2         False  False  False   False  False  False  False
3         False  False  False   False  False  False  False
4         False  False  False   False  False  False  False
..          ...    ...    ...     ...    ...    ...    ...
239       False  False  False   False  False  False  False
240       False  False  False   False  False  False  False
241       False  False  False   False  False  False  False
242       False  False  False   False  False  False  False
243       False  False  False   False  False  False  False

[244 rows x 7 columns]

In [ ]:
df.isna().sum()

total_bill    0
tip           0
sex           0
smoker        0
day           0
time          0
size          0
dtype: int64

Looks like we don't have any missing values in the dataset. Let's say that we had some missing values in the tip column and we figured that they could be filled with the mean tip value, here's how we'd do it.

Filling the null values with the mean of that column is a common practice if the amount of missing data in the column is < 5%. Using the fillna() method to fill the null values, we pass the mean of the column and set the inplace parameter to True (False by default). Inplace will make sure that the mean value is written over the null values and from now on when we call the dataframe, there will be no null values.

In [ ]:
df.fillna(df['tip'].mean(), inplace=True)

Remember the describe() results? Let's confirm that most of the group size's are indeed between 2 and 4 by filtering the values outside that range. 231 out of 244.

In [ ]:
df[(df['size']>1) & (df['size']<5)]

total_bill   tip     sex smoker   day    time  size
0         16.99  1.01  Female     No   Sun  Dinner     2
1         10.34  1.66    Male     No   Sun  Dinner     3
2         21.01  3.50    Male     No   Sun  Dinner     3
3         23.68  3.31    Male     No   Sun  Dinner     2
4         24.59  3.61  Female     No   Sun  Dinner     4
..          ...   ...     ...    ...   ...     ...   ...
239       29.03  5.92    Male     No   Sat  Dinner     3
240       27.18  2.00  Female    Yes   Sat  Dinner     2
241       22.67  2.00    Male    Yes   Sat  Dinner     2
242       17.82  1.75    Male     No   Sat  Dinner     2
243       18.78  3.00  Female     No  Thur  Dinner     2

[231 rows x 7 columns]

In [ ]:
df['total_bill'].rolling(5).sum()

0         NaN
1         NaN
2         NaN
3         NaN
4       96.61
        ...  
239    120.36
240    137.47
241    147.54
242    132.53
243    115.48
Name: total_bill, Length: 244, dtype: float64